### PR TITLE
Configuration parser module mvp

### PR DESCRIPTION
**Pull request by Nicogp**
|Related issue|
|---|
|https://github.com/wazuh/wazuh-agent/issues/26|

## Description
This PR adds the configurationParser module, it reads the configuration from a wazuh.conf file in toml format and then exposes a GetConfig<T>(keys) method to query the different values using one or more keys. 
It uses the TOML11 library, this library is the one that is receiving more updates, it is based on C++20 and supports the latest version of TOML, also its documentation is clear and with examples.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux

- [x] Added unit tests (for new features)

### CODE REVIEW

File: src/agent/configuration_parser/include/toml11.hpp
Comment (TomasTurina) [inline src/agent/configuration_parser/include/toml11.hpp:1]: Why is this file here?

File: src/agent/configuration_parser/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+set(CMAKE_BUILD_TYPE RelWithDebInfo)
+set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -g")
+
+set(CMAKE_TOOLCHAIN_FILE "${CMAKE_SOURCE_DIR}/../../vcpkg/scripts/buildsystems/vcpkg.cmake")
+set(VCPKG_MANIFEST_DIR ${CMAKE_SOURCE_DIR}/../../)
+
+project(ConfigurationParser)
+message(STATUS "******* Project ConfigurationParser *********")
+
+#########################################################################
```
Comment (TomasTurina) [inline src/agent/configuration_parser/CMakeLists.txt:14]: Why is this necessary?

File: src/agent/configuration_parser/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,44 @@
+cmake_minimum_required(VERSION 3.18)
+
+set(CMAKE_CXX_STANDARD 20)
+set(CMAKE_BUILD_TYPE Release)
+set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -g")
```
Comment (jr0me) [inline src/agent/configuration_parser/CMakeLists.txt:6]: We should start using `RelWithDebInfo`.

Also, to my understanding `CMAKE_CXX_FLAGS_DEBUG` only sets flags for `Debug` build types, you can use instead `CMAKE_CXX_FLAGS`.

File: src/agent/configuration_parser/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+#########################################################################
+# Dependencies
+#########################################################################
+
+# Enable testing
+enable_testing()
+
+include_directories(include)
+include_directories(${CMAKE_SOURCE_DIR}/configuration_parser)
```
Comment (jr0me) [inline src/agent/configuration_parser/CMakeLists.txt]: I don't think we should include this folder. Already from line 38 the headers for the configuration parser should be visible to the rest of the project.

File: src/agent/configuration_parser/include/configuration_github.hpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    const bool DEFAULT_ONLY_FEATURE_EVENTS = true;
+    const std::string DEFAULT_EVENT_TYPE = "all";
+
+    class GithubConfigApiAuth
+    {
+        private:
+            std::string orgName;
+            std::string apiToken;
+        public:
+            GithubConfigApiAuth(){};
```
Comment (jr0me) [inline src/agent/configuration_parser/include/configuration_github.hpp]: There's no no need for a semicolon after the closing } of a function definition or a control structure.

File: src/agent/configuration_parser/include/configuration_parser.hpp [hunk 5]
```diff
@@ -16,74 +16,13 @@ namespace configuration
         toml::value tbl;
         configurationGithub::ConfigurationGithubModule configGithub;
 
-        void LoadGithubConfig()
```
Comment (jr0me) [inline src/agent/configuration_parser/include/configuration_parser.hpp]: Since we are separating header and source file in this commit, could we do it in the first commit these files are presented? This will lead to a cleaner git history.

File: src/agent/configuration_parser/include/configuration_github.hpp
Comment (jr0me) [inline src/agent/configuration_parser/include/configuration_github.hpp:1]: Since this file is not needed and was introduced by this branch in the first place, maybe we can avoid adding it altogether.

Comment (jr0me) [reply src/agent/configuration_parser/include/configuration_github.hpp:1]: (Same goes for the rest of the code being removed)

File: src/agent/configuration_parser/CMakeLists.txt [hunk 6]
```diff
@@ -28,18 +32,27 @@ MACRO_ENSURE_OUT_OF_SOURCE_BUILD(
...
-
 include_directories(include)
 include_directories(${CMAKE_SOURCE_DIR})
 
-# Add file to src code
-add_executable(ConfigurationParser_test
-    tests/configuration_parser_test.cpp
-    src/configuration_parser.cpp
-)
+find_package(GTest CONFIG REQUIRED)
```
Comment (jr0me) [inline src/agent/configuration_parser/CMakeLists.txt:38]: This package is only needed when building tests, it should be moved inside the if block below.

File: src/agent/configuration_parser/tests/configuration_parser_test.cpp [hunk 7]
```diff
@@ -1,54 +1,144 @@
...
-    {
-        auto agent_map = parser->GetConfig<std::map<std::string, std::string>>("bad_map");
-    }
-    catch(const std::exception& e)
-    {
-        std::cout << "Exception caught" << '\n';
-    }
-
-    return 0;
+using namespace configuration;
```
Comment (jr0me) [inline src/agent/configuration_parser/tests/configuration_parser_test.cpp:6]: It would be better to introduce the tests from the beginning instead of creating the `main()` function.

File: src/agent/configuration_parser/include/toml11.hpp
Comment (jr0me) [reply src/agent/configuration_parser/include/toml11.hpp:1]: It's been now removed, now toml11 is picked up using Vcpkg.

File: src/agent/configuration_parser/include/configuration_parser.hpp [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+#include <toml.hpp>
+
+#include <iostream>
+
+namespace configuration
+{
+    class ConfigurationParser
+    {
+    private:
+        std::string configFileName = "wazuh.conf";
```
Comment (TomasTurina) [inline src/agent/configuration_parser/include/configuration_parser.hpp:12]: This needs to be modified once the directory structure is defined.

Comment (TomasTurina) [reply src/agent/configuration_parser/include/configuration_parser.hpp:12]: Create a define for `wazuh.conf`.

File: src/agent/configuration_parser/include/configuration_parser.hpp [hunk 9]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <toml.hpp>
+
+#include <iostream>
+
+namespace configuration
+{
+    class ConfigurationParser
```
Comment (TomasTurina) [inline src/agent/configuration_parser/include/configuration_parser.hpp:9]: I think this should be instantiated in the Agent class.

File: src/agent/configuration_parser/include/configuration_parser.hpp [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        auto GetConfig(Ks... ks)
+        {
+            try
+            {
+                auto config = toml::find<T>(tbl, ks...);
+                return config;
+            }
+            catch (const std::exception& e)
+            {
+                std::cout << "The requested value could not be obtained: " << e.what() << std::endl;
```
Comment (TomasTurina) [inline src/agent/configuration_parser/include/configuration_parser.hpp:30]: This should be replaced by a logger once defined.

Comment (jr0me) [reply src/agent/configuration_parser/include/configuration_parser.hpp:30]: Should we create a story to create such logger?

File: src/agent/configuration_parser/include/configuration_parser.hpp (hunk 9 above)
Comment (jr0me) [reply src/agent/configuration_parser/include/configuration_parser.hpp:11]: Done

File: src/agent/configuration_parser/include/configuration_parser.hpp (hunk 8 above)
Comment (jr0me) [reply src/agent/configuration_parser/include/configuration_parser.hpp]: I created a constant hidden in the source file for the configuration parser. Since it was not parameterized and used only once in the default constructor there's no need to store it as a class instance member.

**Review by TomasTurina — APPROVED** [review]
